docs: guide for the built-in control flow (#52369)
angular/angular: aio/content/guide/control_flow.md

# Built-in control flow

Angular templates support *control flow blocks* that let you conditionally show, hide, and repeat elements. 

<div class="alert is-important">

Angular built-in control flow is in [developer preview](/guide/releases#developer-preview). It is ready to try, but may change before becoming stable.

</div>

## `@if` block conditionals

The `@if` block conditionally displays its content when its condition expression is truthy:

```html
@if (a > b) {
   {{a}} is greater than {{b}}
}
```

The `@if` block might have one or more associated `@else` blocks. Immediately after an `@if` block , you can optionally specify any number of `@else if` blocks and one `@else` block:

```html
@if (a > b) {
  {{a}} is greater than {{b}}
} @else if (b > a) {
  {{a}} is less than {{b}}
} @else {
  {{a}} is equal to {{b}}
}
```

### Referencing the conditional expression's result

> The new built in `@if` conditional supports referencing of expression results to keep a solution for common codding patterns:

Reviewer: ```suggestion
The new built-in `@if` conditional supports referencing of expression results to keep a solution for common coding patterns:
```

These variables are always available with these names, but can be aliased via a `let` segment:

```html
@for (item of items; track item.id; let idx = $index, e = $even) {
  Item #{{ idx }}: {{ item.name }}
}
```

The aliasing is especially useful in case of using nested `@for` blocks where contextual variable names could collide. 

### `empty` block

You can optionally include an `@empty` section immediately after the `@for` block content. The content of the `@empty` block displays when there are no items:

```html
@for (item of items; track item.name) {
  <li> {{ item.name }} </li>
} @empty {
  <li> There are no items. </li>
}
```

## `@switch` block - selection

The syntax for `switch` is very similar to `if`, and is inspired by the JavaScript `switch` statement:

```html
@switch (condition) {
  @case (caseA) {
    Case A.
  }
  @case (caseB) {
    Case B.
  }
  @default {
    Default case.
  }
}
```

The value of the conditional expression is compared to the case expression using the `===` operator. 

**`@switch` does not have fallthrough**, so you do not need an equivalent to a `break` or `return` statement.

The `@default` block is optional and can be omitted.

> ## Built in control flow and the `NgIf`, `NgSwitch` and `NgFor` structural directives

Reviewer: ```suggestion
## Built-in control flow and the `NgIf`, `NgSwitch` and `NgFor` structural directives
```

The `@if` block replaces `*ngIf` for expressing conditional parts of the UI.

> The `@switch` block replaces `ngSwitch` with major benefits:

Reviewer 2: ```suggestion
The `@switch` block replaces `*ngSwitch` with major benefits:
```

Author: `ngSwitch` does _not_ require *, see https://angular.io/api/common/NgSwitch

* it does not require a container element to hold the condition expression or each conditional template;

> * it can support template type-checking, including type narrowing within each branch.

Author: Yeh, lol

> The for block replaces `*ngFor` for iteration, and has several differences compared to its structural directive `NgFor` predecessor:

Reviewer: ```suggestion
The `@for` block replaces `*ngFor` for iteration, and has several differences compared to its structural directive `NgFor` predecessor:
```

> # Built-in control flow

Author: Right, those are regular Angular expressions - what is allowed in a template is legal here. I wonder if we should highlight it since we are not mentioning it explicitly in all the other Angular features. I would rather skip it but don't feel too strongly about it.

> In the Angular's initial design control flow was expressed via composition of user-facing concepts: a structural directive on an `<ng-template>` element would determine when and where embedded views are created from that template. To provide authoring experience familiar to UI developers the * microsyntax was added to the template language.

Author: Got rid of the historical context and moved discussion on  NgIf, NgFor etc to a separate section.

> Many templating languages  have control flow primitives and Angular introduces one in version 17. Our review of developer experience pain points highlighted microsyntax-based control flow as having significant weaknesses. The proposed built-in control flow syntax addresses these issues and significantly improves the developer experience, in addition to being a foundation for new features.

Author: Dropped.

<div class="alert is-important">

Angular built-in control flow is in [developer preview](/guide/releases#developer-preview): it is ready for you to try, but may change before becoming stable.

</div>

> ## `if` block conditionals

Author: Works for me!

The `if` block replaces `*ngIf` for expressing conditional parts of the UI.

The primary if block includes an expression representing the condition under which this block will be displayed:

```html
@if (a > b) {
   {{a}} is greater than {{b}}
}
```

The if block groups may also contain one or more else blocks. One bare else block is allowed, as well as zero or more else blocks with an additional condition (indicated by the `if` keyword following the `else` name):

```html
@if (a > b) {
  {{a}} is greater than {{b}}
} @else if (b > a) {
  {{a}} is less than {{b}}
} @else {
  {{a}} is equal to {{b}}
}
```

> ### aliasing the conditional expression

Author: Makes sense, thnx!

The current `NgIf` directive supports aliasing of the condition expression via `as`:

```html
<div *ngIf="users$ | async as users">
  {{ users.length }}
</div>
```

To enable direct automatic migration from `*ngIf` to the new built in conditional statement, aliasing will be supported in `@if``:

```html
@if (users$ | async; as users) {
  {{ users.length }}
}
```

> ## `for` block - repeaters

Author: Done.

> The for block replaces `*ngFor` for iteration, and has several differences compared to its structural directive predecessor. A basic for block group looks like:

Author: Went with a slightly different wording.

```html
@for (item of items; track item.id) {
  {{ item.name }}
}
```

### `track` for calculating difference of two collections

The `track` setting replaces `NgFor`'s concept of a `trackBy` function. It determines the row key which for will use to associate array items with the views it creates, moving them around as needed. Because `for` is built-in, we can provide a better experience than passing a `trackBy` function, and directly use an expression representing the key instead.

Migrating from trackBy to track is possible by invoking the `trackBy` function:

```html
@for (item of items; track itemId($index, item)) {
  {{ item.name }}
}
```

> In our performance research, we've identified `NgFor` loops over immutable data without `trackBy` as one of the most common causes for performance issues across Angular applications. Because of the potential for poor performance, we’re making `track` required for `for` loops. When in doubt, using `track $index` is a good default.

Reviewer 3: `In our performance research, we've identified `NgFor` loops over immutable data without `trackBy` as one of the most common causes for performance issues across Angular applications. Because of the potential for poor performance, we’re making `track` required for `for` loops. When in doubt, using `track $index` is a good default.`

If that's the case, why do I have to write it in the first place? Not the place to debate issues, but I feel like if you have to word something like this, it points to a deeper design problem.

https://github.com/angular/angular/issues/52050#issuecomment-1776414729

> In our performance research, we've identified `NgFor` loops over immutable data without `trackBy` as one of the most common causes for performance issues across Angular applications. Because of the potential for poor performance, we’re making `track` required for `for` loops. When in doubt, using `track $index` is a good default.

Author: @jpike88 let's keep this discussion in https://github.com/angular/angular/issues/52050 - I've re-opened it so we can capture all the feedback.

> In our performance research, we've identified `NgFor` loops over immutable data without `trackBy` as one of the most common causes for performance issues across Angular applications. Because of the potential for poor performance, we’re making `track` required for `for` loops. When in doubt, using `track $index` is a good default.

Author: @jelbourn reworded.

### `$index` and other contextual variables

Within `for` row views, there are several implicit variables which are always available:

| Variable | Meaning |
| -------- | ------- |
| `$count` | Number of items in a collection iterated over |
| `$index` | Index of the current row |
| `$first` | Whether the current row is the first row |
| `$last` | Whether the current row is the last row |
| `$even` | Whether the current row index is even |
| `$odd` | Whether the current row index is odd |

These variables are always available with these names, but can be aliased via a `let` segment:

```html
@for (item of items; track item.id; let idx = $index, e = $even) {
  Item #{{ idx }}: {{ item.name }}
}

> ```

Author: Mentioned.

### `empty` block

The `for` block supports of providing an additional template to render when there are no items in a list: 

```html
@for (item of items; track item.name) {
  <li> {{ item.name }} </li>
} @empty {
  <li> There are no items. </li>
}
```

### `IterableDiffer`

Unlike its structural directive predecessor, the new `for` will not use Angular’s customizable diffing implementation (`IterableDiffer`), but instead will use a new optimized algorithm for calculating a minimal number of DOM operations to be performed in response to changes in a collection.

> ## `switch` block - selection

Author: Good call - added!

The syntax for `switch` is very similar to `if`, and is inspired by the JavaScript `switch` statement:

```html
@switch (condition) {
  @case (caseA) {
    Case A.
  }
  @case (caseB) {
    Case B.
  }
  @default {
    Default case.

>   }

Author: Good call, added!

}
```

The `switch` built-in has several major benefits over `NgSwitch`:

* it does not require a container element to hold the condition expression or each conditional template;
* it can support template type-checking, including type narrowing within each branch.

> `switch` does not have fallthrough (there is no `break` operation required).

Reviewer 4: ```suggestion
**`@switch` does not have fallthrough**, so you do not need an equivalent to a `break` or `return` statement.
```

# Built-in control flow

> In the Angular's initial design control flow was expressed via composition of user-facing concepts: a structural directive on an `<ng-template>` element would determines when and where embedded views are created from that template. To provide authoring experience familiar to UI developers the * microsyntax was added to the template language.

Reviewer 5: "would determines" -> "would determine"

> Many templating languages  have control flow primitives and Angular introduces one in version 17. Our review of developer experience pain points highlighted microsyntax-based control flow as having significant weaknesses. The proposed built-in control flow syntax addresses these issues and significantly improves the developer experience, in addition to being a foundation for new features.

Reviewer 5: "Angular introduces one" -> "Angular is introducing one"

<div class="alert is-important">

Angular built-in control flow is in [developer preview](/guide/releases#developer-preview): it is ready for you to try, but may change before becoming stable.

</div>

## The block syntax

The primary syntactic element of the new control flow design is a named block group composed of one or more named blocks. Block groups are template nodes, occupying the same syntactic positions as elements, text nodes, and text bindings. Each block in a block group contains Angular template syntax within it, as a list of child template nodes (potentially including additional block groups).

> Each block group begins with the `@` character followed by a block name and the opening curly brace (`{`). A bock ends with a matching closing curly brace (`}`). Tags can contain specific syntax based on their name. As an example, the basic `if` control flow structure looks like: 

Reviewer 5: Should we mention parentheses around the arguments?

> Each block group begins with the `@` character followed by a block name and the opening curly brace (`{`). A bock ends with a matching closing curly brace (`}`). Tags can contain specific syntax based on their name. As an example, the basic `if` control flow structure looks like: 

Reviewer 5: "A bock" -> "A block"

> Each block group begins with the `@` character followed by a block name and the opening curly brace (`{`). A bock ends with a matching closing curly brace (`}`). Tags can contain specific syntax based on their name. As an example, the basic `if` control flow structure looks like: 

Reviewer 6: ```suggestion
Each block group begins with the `@` character followed by a block name and the opening curly brace (`{`). A block ends with a matching closing curly brace (`}`). Tags can contain specific syntax based on their name. As an example, the basic `if` control flow structure looks like: 
```

Author: Done.

```html

> @if cond.expr {

Reviewer 5: Needs parentheses around the expression.

Author: Ohhh, I though that we can figure it out from parsing the full expression

Author: Changing

  <child-cmp />
}
```

### Tag syntax

Each named block group can customize the tag syntax following the tag name and a space. For example, the if block group supports an expression following the if, which is used as the if condition.

Different block groups may use this syntactic space in different ways.

### Related blocks

In addition to the primary named block created by the block group itself, additional blocks may be defined in relation to it. Related blocks follow a set of syntactic rules:

> * valid names for these blocks are statically determined by the overall block group to which they belong. Block names are always part of a statically known set and are never user-determined;

Reviewer 5: Seems like these bullet points should start with a capital letter and end with a full stop.

Author: Hmm, actually it would seem strange to me to start with an uppercase after `:`. Let's leave as-is?

* additional blocks may be either required or optional, depending on the block group, its options, or the presence/absence of other named blocks;
* multiple blocks of the same name are allowed;
* the order of blocks in a block group may be semantically meaningful, or may be fixed by the type of block group.

For example, an if block group may define an optional else block:

```html

> @if cond.exp) {

Reviewer 5: Syntax error here.

> @if cond.exp) {

Reviewer 6: ```suggestion
@if (cond.exp) {
```

    <true-cmp />
} @else {
   <false-cmp />
}
```

> The above syntax creates an if block group with two blocks: the primary `if`` block (containing the `<true-cmp>`) and an else block (containing the `<false-cmp>`).

Reviewer 5: Looks like there's an extra backtick after `if`.

### Optional parentheses

> In JavaScript, control flow statements require parentheses (`if (cond)``). Because block tags are wrapped in braces already, they don't require parentheses. However, we support including parentheses if desired:

Reviewer 5: Extra backtick before the `)`.

> In JavaScript, control flow statements require parentheses (`if (cond)``). Because block tags are wrapped in braces already, they don't require parentheses. However, we support including parentheses if desired:

Reviewer 5: > Because block tags are wrapped in braces already, they don't require parentheses

I'm not sure I understand this part. The parentheses are only optional if you don't have arguments for the block.

```html
@if (cond.expr) {
  <child-cmp />
}
```

> If used, the parentheses wrap the entire body of the block tag.

Reviewer 5: Is this referring to braces, not parentheses?

## `if` block conditionals

> The `if`` block replaces `*ngIf` for expressing conditional parts of the UI.

Reviewer 5: Extra backtick before "block"

Author: Uhh, this is annoying, blaming it all on the editor :-)

The primary if block includes an expression representing the condition under which this block will be displayed:

```html
@if (a > b) {
   {{a}} is greater than {{b}}
}
```

> The if block groups may also contain one or more else blocks. One bare else block is allowed, as well as zero or more else blocks with an additional condition (indicated by the `if` keyword following the `else` name):

Reviewer 5: I'm not sure I understand what "bare else block" means in this case.

Author: This got changed.

```html
@if (a > b) {
  {{a}} is greater than {{b}}
} @else if (b > a) {
  {{a}} is less than {{b}}
} @else {
  {{a}} is equal to {{b}}
}
```

### aliasing the conditional expression

The current `NgIf` directive supports aliasing of the condition expression via `as`:

```html
<div *ngIf="users$ | async as users">
  {{ users.length }}
</div>
```

> To enable direct automatic migration from `*ngIf`` to the new built in conditional statement, aliasing will be supported in `@if``:

Reviewer 5: Extra backtick at ngIf.

```html
@if (users$ | async; as users) {
  {{ users.length }}
}
```

## `for` block - repeaters

> The for block replaces `*ngFor`` for iteration, and has several differences compared to its structural directive predecessor. A basic for block group looks like:

Reviewer 5: Extra backtick after ngFor.

```html
@for (item of items; track item.id) {
  {{ item.name }}
}
```

### `track` for calculating difference of two collections

> The `track`` setting replaces `NgFor`'s concept of a `trackBy` function. It determines the row key which for will use to associate array items with the views it creates, moving them around as needed. Because `for`` is built-in, we can provide a better experience than passing a `trackBy` function, and directly use an expression representing the key instead.

Reviewer 5: Extra backtick after track. Also one after "Because for".

Migrating from trackBy to track is possible by invoking the `trackBy` function:

```html
@for (item of items; track itemId($index, item)) {
  {{ item.name }}
}
```

> In our performance research, we've identified `NgFor` loops over immutable data without `trackBy` as one of the most common causes for performance issues across Angular applications. Because of the potential for poor performance, we’re making `track` required for `for` loops. When it doubt, using `track $index` is a good default.

Reviewer 6: ```suggestion
In our performance research, we've identified `NgFor` loops over immutable data without `trackBy` as one of the most common causes for performance issues across Angular applications. Because of the potential for poor performance, we’re making `track` required for `for` loops. When in doubt, using `track $index` is a good default.
```

### `$index`` and other contextual variables

Within `for` row views, there are several implicit variables which are always available:

> | Variable | Meaning |

Reviewer 5: The table is missing `$count`.

Author: Good catch!

| -------- | ------- |
| `$index` | Index of the current row |
| `$first` | Whether the current row is the first row |
| `$last` | Whether the current row is the last row |
| `$even` | Whether the current row index is even |
| `$odd` | Whether the current row index is odd |

These variables are always available with these names, but can be aliased via a `let` segment:

```html

> @for item of items; track item.id; let idx = $index, e = $even {

Reviewer 5: Missing parentheses.

  Item #{{ idx }}: {{ item.name }}
}
```

### `empty` block

The `for` block supports of providing an additional template to render when there are no items in a list: 

```html

> @for item of items; track item.name {

Reviewer 5: Missing parentheses here as well.

>   <li> {{ item.name }}

Reviewer 6: ```suggestion
  <li> {{ item.name }} </li>
```

} @empty {

>   <li> There are no items.

Reviewer 6: ```suggestion
  <li> There are no items. </li>
```
